### PR TITLE
Add handling to process for adding InOut to ShipperTransportation when InOut has no HUs

### PR DESCRIPTION
**Pull request by TheBestPessimist**
https://github.com/metasfresh/metasfresh/issues/5885

### CODE REVIEW

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/InOutPackageDAO.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metasfresh/metasfresh/pullRequest?prid=4699667&bid=15704934)

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/InOutPackageDAO.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metasfresh/metasfresh/pullRequest?prid=4699667&bid=15704934)

File: de.metas.handlingunits.base/src/main/java/de/metas/picking/service/impl/HU2PackingItemsAllocator.java [hunk 1]
```diff
@@ -519,7 +519,7 @@ private void pickFromVHU(@NonNull final I_M_HU pickFromVHU, @NonNull final Packi
 		final StockQtyAndUOMQty stockQtyAndUomQty = CatchWeightHelper.extractQtys(_huContext, getProductId(), qtyPacked, pickFromVHU);
 
 		// "Back" allocate the qtyPicked from VHU to given shipment schedule
-		huShipmentScheduleBL.addQtyPickedAndUpdateHU(shipmentSchedule, stockQtyAndUomQty, pickFromVHU, _huContext);
+		huShipmentScheduleBL.addQtyPickedAndUpdateHU(shipmentSchedule, stockQtyAndUomQty, pickFromVHU, _huContext, false);
```
Comment (metas-ts) [inline de.metas.handlingunits.base/src/main/java/de/metas/picking/service/impl/HU2PackingItemsAllocator.java]: follow-up:
```suggestion
		huShipmentScheduleBL.addQtyPickedAndUpdateHU(shipmentSchedule, stockQtyAndUomQty, pickFromVHU, _huContext, false/*IsAnonyouslyPickedOnTheFly*/);
```
not everyone has intelliJ displaying the parameter names nicely :-)

File: de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/shipmentschedule/api/impl/ShipmentScheduleHUTrxListener.java [hunk 2]
```diff
@@ -84,7 +84,7 @@ public void trxLineProcessed(final IHUContext huContext, final I_M_HU_Trx_Line t
 		//
 		// Link VHU to shipment schedule
 		final IHUShipmentScheduleBL huShipmentScheduleBL = Services.get(IHUShipmentScheduleBL.class);
-		huShipmentScheduleBL.addQtyPickedAndUpdateHU(shipmentSchedule, stockQtyAndUomQty, vhu, huContext);
+		huShipmentScheduleBL.addQtyPickedAndUpdateHU(shipmentSchedule, stockQtyAndUomQty, vhu, huContext, false);
```
Comment (metas-ts) [inline de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/shipmentschedule/api/impl/ShipmentScheduleHUTrxListener.java]:
```suggestion
		huShipmentScheduleBL.addQtyPickedAndUpdateHU(shipmentSchedule, stockQtyAndUomQty, vhu, huContext, false/*IsAnonyouslyPickedOnTheFly*/);
```

File: de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/shipmentschedule/api/IHUShipmentScheduleBL.java [hunk 3]
```diff
@@ -37,9 +37,10 @@
 	 * And finally update the given {@code tuOrVHU}'s status to "Picked".
 	 *
 	 * @param tuOrVHU TU or VirtualHU to link on
+	 * @param anonymousTuPickedOnTheFly
```
Comment (metas-ts) [inline de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/shipmentschedule/api/IHUShipmentScheduleBL.java]: follow-up: avoid lying or useless javadoc

File: de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/picking/candidate/commands/ProcessPickingCandidatesCommand.java [hunk 4]
```diff
@@ -218,7 +218,8 @@ private void addShipmentScheduleQtyPickedAndUpdateHU(
 						qtyPicked,
 						hu),
 				hu,
-				huContext);
+				huContext,
+				false);
```
Comment (metas-ts) [inline de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/picking/candidate/commands/ProcessPickingCandidatesCommand.java]:
```suggestion
				false/*IsAnonyouslyPickedOnTheFly*/);
```

File: de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HandlingUnitsBL.java [hunk 5]
```diff
@@ -518,6 +519,25 @@ public boolean isTopLevel(final I_M_HU hu)
 		return topLevel;
 	}
 
+	@Override
+	public boolean isAnonymousHuPickedOnTheFly(@NonNull final I_M_HU hu, final ImmutableList<InOutLineId> lineIds)
```
Comment (metas-ts) [inline de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HandlingUnitsBL.java]: follow-up: not sure why we need `lineIds` as parameter?
if an HU is anonymousHuPickedOnTheFly, but for an unrelated inoutLine..how could the happen, and: you you then want to consider it *not* as anonymously picked?

File: de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java [hunk 6]
```diff
@@ -140,6 +140,54 @@
...
+	@NonNull
+	@Override
+	public ImmutableList<I_M_Package> addInOutWithoutHUToShipperTransportation(@NonNull final ShipperTransportationId shipperTransportationId, @NonNull final ImmutableList<de.metas.inout.model.I_M_InOut> inOuts)
+	{
+		final I_M_ShipperTransportation shipperTransportation = load(shipperTransportationId, I_M_ShipperTransportation.class);
+
+		// Make sure Shipper Transportation is still open
+		if (shipperTransportation.isProcessed())
+		{
+			throw new AdempiereException("@M_ShipperTransportation_ID@: @Processed@=@Y@");
```
[the remark this answers is not in the document]
Comment (metas-ts) [reply de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java]: not yet done; see my (upcoming) comment

File: de.metas.swat/de.metas.swat.base/src/main/java-gen/de/metas/inoutcandidate/model/I_M_ShipmentSchedule_QtyPicked.java [hunk 7]
```diff
@@ -157,6 +157,29 @@
...
+	 * <br>Type: YesNo
+	 * <br>Mandatory: true
+	 * <br>Virtual Column: false
+	 */
+	public boolean isAnonymousHuPickedOnTheFly();
+
+    /** Column definition for isAnonymousHuPickedOnTheFly */
+    public static final org.adempiere.model.ModelColumn<I_M_ShipmentSchedule_QtyPicked, Object> COLUMN_isAnonymousHuPickedOnTheFly = new org.adempiere.model.ModelColumn<I_M_ShipmentSchedule_QtyPicked, Object>(I_M_ShipmentSchedule_QtyPicked.class, "isAnonymousHuPickedOnTheFly", null);
+    /** Column name isAnonymousHuPickedOnTheFly */
+    public static final String COLUMNNAME_isAnonymousHuPickedOnTheFly = "isAnonymousHuPickedOnTheFly";
```
[the remark this answers is not in the document]
Comment (metas-ts) [reply de.metas.swat/de.metas.swat.base/src/main/java-gen/de/metas/inoutcandidate/model/I_M_ShipmentSchedule_QtyPicked.java]: also rename it in the java model class - not just in the application dictionary

[the remark this answers is not in the document]
Comment (TheBestPessimist) [reply de.metas.swat/de.metas.swat.base/src/main/java-gen/de/metas/inoutcandidate/model/I_M_ShipmentSchedule_QtyPicked.java]: also renamed the class

File: de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (TheBestPessimist) [reply de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/impl/HUShipperTransportationBL.java]: solved now

**Pull request merged**